main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -17,29 +17,284 @@
+    * required for correctness.
+    * */
+  private def moveErrEquations(state: State): Unit = {
+    state.eqs = state.eqs.filter(eq => {
+      if (eq.isPending) {
+        true
+      } else {
+        state.errEqs = eq :: state.errEqs
+        false
+      }
+    })
+  }
+
+  /**
+    * Attempts to solve the equation system `eqs`.
+    *
+    * All returned equations are either marked [[Equation.Status.Unsolvable]] or
+    * [[Equation.Status.Timeout]].
+    */
+  def solve(eqs: List[Equation])(implicit opts: Options): (List[Equation], SetSubstitution) = {
+    val (solvedEqs, s, _) = solveWithInfo(eqs)
+    (solvedEqs, s)
+  }
+
+  /**
+    * Attempts to solve the equation system `eqs`, like [[solve]].
+    *
+    * Additionally returns the name and number of the last phase to make progress.
     */
-  private final case class Options(sveRecSizeThreshold: Int)
+  def solveWithInfo(l: List[Equation])(implicit opts: Options): (List[Equation], SetSubstitution, (String, Int)) = {
+    val noDebug = opts.copy(debugging = Options.Debugging.Nothing)
 
-  /** A class to track progress during set unification. */
-  private final class Progress() {
+    val state = new State(l)
 
-    /** Is `true` if [[markProgress]] has been called. */
-    private var hasMadeProgress: Boolean = false
+    runPhase(runRule(constantAssignment), state, "Constant Assignment", "Solves equations where one or both sides are ground.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(runRule(variableAlias), state, "Variable Aliases", "Solves equations like `x1 ~ x2`.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(runRule(variableAssignment), state, "Simple Variable Assignment", "Solves non-recursive variable assignments.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(duplicatedAndReflective, state, "Duplicates and Reflective", "Solves `f ~ f` and duplicated formulas.")
+    runPhase(runRule(trivial), state, "Trivial Equations", "Solves trivial equations.")(noDebug)
+    runPhase(assertEquationCount, state, "Assert Size", "Quits if there are too many equations.")(noDebug)
+    runPhase(svePermuations, state, "SVE", "Applies SVE in different permutations.")
 
-    /** Mark that progress has been made. */
-    def markProgress(): Unit = hasMadeProgress = true
+    assert(state.eqs.isEmpty, message = state.eqs)
+
+    (state.allEqs, state.subst, (state.lastProgressPhaseName, state.lastProgressPhaseNumber))
+  }
 
-    /** Returns `true` if [[markProgress]] has been called. */
-    def hasProgressed: Boolean = hasMadeProgress
+  /** Marks all equations as [[Equation.Status.Timeout]] if there are more than [[Options.sizeThreshold]]. */
+  private def assertEquationCount(eqs: List[Equation])(implicit opts: Options): Option[(List[Equation], SetSubstitution)] = {
```

Comment:
I am confused about what this function does and its documentation.

I dont think I would use the terminolog "complex equations in subst."

Reply:
I've updated it now. Maybe we just remove it since we also have `sveRecSizeThreshold` which more precisely controls the same issue, stopping SVE from effectively running forever?